device/transport/mqtt/src/mqtt.ts

```diff
@@ -547,4 +608,222 @@ export class Mqtt extends EventEmitter implements Client.Transport, StableConnec
 
     this._fsm.handle('getTwinReceiver', done);
   }
+
+  onDeviceMethod(methodName: string, callback: (methodRequest: DeviceMethodRequest, methodResponse: DeviceMethodResponse) => void): void {
+    this.on('method_' + methodName, callback);
+  }
+
+  private _setupSubscription(topic: TopicDescription): void {
+    debug('subscribe: ' + JSON.stringify(topic));
+    topic.subscribeInProgress = true;
+    this._mqtt.subscribe(topic.name, { qos: 0 }, (err) => {
+      topic.subscribeInProgress = false;
+      if (!err) {
+        topic.subscribed = true;
+      }
+
+      // TODO: There doesn't seem to be a way in the current design to surface
+      //       the error if the MQTT topic subscription fails. Fix this.
+    });
+  }
+
+  private _removeSubscription(topic: TopicDescription): void {
+    debug('unsubscribe ' + JSON.stringify(topic));
+    this._mqtt.unsubscribe(topic.name, (err) => {
+      if (!err) {
+        topic.subscribed = false;
+
+        // TODO: There doesn't seem to be a way in the current design to surface
+        //       the error if the MQTT topic unsubscription fails. Fix this.
+      }
+    });
+  }
+
+  private _dispatchMqttMessage(topic: string, payload: any): void {
+    debug('message received on ' + topic);
+    debug(JSON.stringify(payload ? payload.toString() : null));
+    // dispatch the message to the appropriate handler
+    let targetTopic = null;
+    Object.keys(this._topics).some((topicIndex) => {
+      // Turns out regexes are stateful. We need to reset the search index back to
+      // the beginning every time we use it.
+      const theTopic = this._topics[topicIndex];
+      theTopic.topicMatchRegex.lastIndex = 0;
+      if (theTopic.topicMatchRegex.test(topic)) {
+        targetTopic = theTopic;
+      }
+      return targetTopic !== null;
+    });
+    if (!!targetTopic) {
+      targetTopic.handler(topic, payload);
```

Comment:
The claim would be that this is self documenting code and that I should be able to understand exactly what is going on here.

It isn't, I don't.

Don't expect a change but, hey, descriptions about how something works conceptually are a good thing.

Reply:
added a few comments

---
In reply to: [141920173](https://github.com/Azure/azure-iot-sdk-node/pull/96#discussion_r141920173) [](ancestors = 141920173)